CRM/Contribute/BAO/Contribution.php

```diff
@@ -1722,8 +1722,24 @@ public static function transitionComponents($params, $processContributionObject
       if (is_array($memberships)) {
         foreach ($memberships as $membership) {
           if ($membership) {
-            $membership->status_id = array_search('Cancelled', $membershipStatuses);
+            $newStatus = array_search('Cancelled', $membershipStatuses);
+
+            // Create activity
+            $allStatus = CRM_Member_BAO_Membership::buildOptions('status_id', 'get');
+            $activityParam = array(
+              'subject' => "Status changed from {$allStatus[$membership->status_id]} to {$allStatus[$newStatus]}",
+              'source_contact_id' => CRM_Core_Session::singleton()->get('userID'),
+              'target_contact_id' => $membership->contact_id,
+              'source_record_id' => $membership->id,
+              'activity_type_id' => CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_Activity', 'activity_type_id', 'Change Membership Status'),
```

Comment:
for these rows you don't need to resolve the values ie

'activity_type_id' => 'Change Membership Status',
'status_id' => 'Completed',
'priority_id' => 'Normal',
'activity_date_time' =>'now', 

some of those might be the defaults & not need setting.

The api handles those & they are more readable and it removes the hard-coding.

Note I have only looked at this from a code POV - not checked there is agreement on the problem.

What I would REALLY like to do is help you to write a unit test for this.